kpatch-build/lookup.c

```diff
@@ -223,6 +236,132 @@ int lookup_is_exported_symbol(struct lookup_table *table, char *name)
 	return 0;
 }
 
+static void *grab_file(const char *filename, unsigned long *size) {
+	struct stat st;
+	void *map = MAP_FAILED;
+	int fd;
+
+	fd = open(filename, O_RDONLY);
+	if (fd < 0)
+		return NULL;
+	if (fstat(fd, &st))
+		goto failed;
+
+	*size = st.st_size;
+	map = mmap(NULL, *size, PROT_READ|PROT_WRITE, MAP_PRIVATE, fd, 0);
+
+failed:
+	close(fd);
+	if (map == MAP_FAILED)
+		return NULL;
+	return map;
+}
+
+/**
+  * Return a copy of the next line in a mmap'ed file.
+  * spaces in the beginning of the line is trimmed away.
+  * Return a pointer to a static buffer.
+  **/
+static char *get_next_line(unsigned long *pos, void *file, unsigned long size)
+{
+	static char line[4096];
+	int skip = 1;
+	size_t len = 0;
+	signed char *p = (signed char *)file + *pos;
+	char *s = line;
+
+	for (; *pos < size ; (*pos)++) {
+		if (skip && isspace(*p)) {
+			p++;
+			continue;
+		}
+		skip = 0;
+		if (*p != '\n' && (*pos < size)) {
+			len++;
+			*s++ = *p++;
+			if (len > 4095)
+				break; /* Too long, stop */
+		} else {
+			/* End of string */
+			*s = '\0';
+			return line;
+		}
+	}
+	/* End of buffer */
+	return NULL;
+}
+
+static void release_file(void *file, unsigned long size)
+{
+	munmap(file, size);
+}
+
+/* replace '-' with '_' */
+static void fix_modname(char *modname)
+{
+	char *cur;
+
+	if (!modname)
+		return;
+
+	cur = modname;
+	while (*cur != '\0') {
+		if (*cur == '-')
+			*cur = '_';
+		cur++;
+	}
+}
+
+/*
+ * lookup_exported_symbol_objname - find the object/module an exported
+ * symbol belongs to.
+ *
+ * Based on read_dump() from scripts/mod/modpost.c with a few modifications.
+ *
+ * Module.symvers line format:
+ * 0x12345678<tab>symbol<tab>module[[<tab>export]<tab>something]
+ */
+char *lookup_exported_symbol_objname(const char *fname, char *symname)
```

Comment:
This opens and reads the file every time the function is called.   I think it would be cleaner to read the file once in lookup_open() like my version from a few years ago: https://github.com/jpoimboe/kpatch/commit/8cdca59c8844505bcf40b59b5c5050b03e974fea